lib/pbench/test/unit/agent/task/test_generate_token.py

```diff
@@ -0,0 +1,167 @@
+from click.testing import CliRunner
+import requests
+import responses
+
+from pbench.cli.agent.commands import generate_token
+
+
+class TestGenerateToken:
+
+    USER_SWITCH = "--username"
+    PSWD_SWITCH = "--password"
+    USER_PROMPT = "Username: "
+    PSWD_PROMPT = "Password: "
+    USER_TEXT = "test_user"
+    PSWD_TEXT = "password"
+    TOKEN_TEXT = "what is a token but 139 characters of gibberish"
+    URL = "http://pbench.example.com/api/v1"
+    ENDPOINT = "/login"
+
+    @staticmethod
+    def add_success_mock_response():
+        responses.add(
+            responses.POST,
+            TestGenerateToken.URL + TestGenerateToken.ENDPOINT,
+            status=200,
+            json={"auth_token": TestGenerateToken.TOKEN_TEXT},
+        )
+
+    @staticmethod
+    def add_badlogin_mock_response():
+        responses.add(
+            responses.POST,
+            TestGenerateToken.URL + TestGenerateToken.ENDPOINT,
+            status=403,
+            json={"message": "Bad login"},
+        )
+
+    @staticmethod
+    def add_connectionerr_mock_response():
+        responses.add(
+            responses.POST,
+            TestGenerateToken.URL + TestGenerateToken.ENDPOINT,
+            body=requests.exceptions.ConnectionError(
```

Comment:
But no linked `reason` exception?

Reply:
Yeah...I thought about trying to dive into the nuances of the construction of `requests.exceptions.ConnectionError` -- I did give it a reasonable try -- but after awhile it seemed like it was more effort than was required for a mock.

The escapade did, however, lead to the to conclusion that part of the error reporting in the tool code around dealing with exceptions from the request was somewhat fragile (it is more robust, now), so it was worthwhile in that respect.